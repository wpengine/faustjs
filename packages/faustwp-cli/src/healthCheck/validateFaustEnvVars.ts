import { getWpSecret } from '../utils/index.js';
import { errorLog, warnLog } from '../stdout/index.js';

/**
 * Validates that the appropriate Faust related environment variables are set.
 */
export const validateFaustEnvVars = () => {
  if (!process.env.NEXT_PUBLIC_WORDPRESS_URL) {
    errorLog('Could not find NEXT_PUBLIC_WORDPRESS_URL environment variable.');

    process.exit(0);
  }

  if (!getWpSecret()) {
    warnLog('Could not find FAUST_SECRET_KEY environment variable.');
    warnLog('Some functionality may be limited.');
  }

  if (
    process.env.NEXT_PUBLIC_WORDPRESS_URL.startsWith('http://') &&
    getWpSecret()
  ) {
    warnLog('Your WordPress site is not running on https!');
    warnLog(
      'This is a security concern as all traffic with your secret key is in plain text.',
    );
    warnLog(
      'Please make sure your production Faust app runs with a WordPress instance on https!',
    );
  }
};
