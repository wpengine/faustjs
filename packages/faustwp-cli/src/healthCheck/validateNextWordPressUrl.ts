import { getWpSecret, getWpUrl } from '../utils/index.js';
import { errorLog, warnLog } from '../stdout/index.js';

/**
 * Validates the NEXT_PUBLIC_WORDPRESS_URL environment variable by sending a POST request to the Faust Plugin API.
 * If the URL matches the Faust Plugin Headless URL, the validation fails, and an error is logged.
 */
export async function validateNextWordPressUrl(): Promise<void> {
  const apiUrl = `${getWpUrl()}/wp-json/faustwp/v1/validate_public_wordpress_url`;
  const headers = {
    'Content-Type': 'application/json',
    'x-faustwp-secret': getWpSecret() || '',
  };

  const postData = {
    public_wordpress_url: getWpUrl(),
  };
  try {
    const response = await fetch(apiUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify(postData),
    });

    if (!response.ok) {
      if (response.status === 404) {
        // Handle the case when the route does not exist
        warnLog(
          'Route not found: Please update your FaustWP plugin to the latest version.',
        );
      } else {
        errorLog(
          'Validation Failed: Your NEXT_PUBLIC_WORDPRESS_URL value is misconfigured. It should match your WordPress site URL and not your Faust front-end site URL.',
        );
        process.exit(1);
      }
    }
  } catch (error) {
    console.log('error', error);
  }
}
