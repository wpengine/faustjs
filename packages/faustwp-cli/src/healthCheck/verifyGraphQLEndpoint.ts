import 'isomorphic-fetch';

import { errorLog, debugLog } from '../stdout/index.js';
import { getGraphqlEndpoint } from '../utils/index.js';

/**
 * Checks if the WordPress GraphQL endpoint is available to use.
 */
export async function verifyGraphQLEndpoint() {
  const graphqlEndpoint = getGraphqlEndpoint();

  try {
    // Perform GraphQL request.
    const response: Response = await fetch(graphqlEndpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query: '{ __typename }' }),
    });

    // Expect a valid GraphQL response.
    const json: {
      data: {
        __typename: 'string';
      };
    } = await response.json();

    // eslint-disable-next-line no-underscore-dangle
    if (!json.data.__typename) {
      throw new Error(`GraphQL request didn't include "json.data.__typename"`);
    }

    debugLog(`Discoverd GraphQL endpoint at ${graphqlEndpoint}`);

    return true;
  } catch (err) {
    errorLog(`Unable to find a GraphQL endpoint at ${graphqlEndpoint}`);
    errorLog(
      'WPGraphQL may not be active, or your WordPress site is unavailable.',
    );

    return process.exit(0);
  }
}
