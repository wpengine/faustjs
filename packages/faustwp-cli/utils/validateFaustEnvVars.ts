import { errorLog, warnLog } from './log.js';

/**
 * Validates that the appropriate Faust related environment variables are set.
 */
export const validateFaustEnvVars = () => {
  if (!process.env.NEXT_PUBLIC_WORDPRESS_URL) {
    errorLog(
      'Please provide a NEXT_PUBLIC_WORDPRESS_URL environment variable.',
    );

    process.exit(0);
  }

  if (!process.env.FAUSTWP_SECRET_KEY) {
    warnLog(
      'You do not have a headless secret key specified. Some functionality may be limited.',
    );
  }
};
