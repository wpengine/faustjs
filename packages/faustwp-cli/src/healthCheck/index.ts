import { validateFaustEnvVars } from './validateFaustEnvVars.js';
import { verifyGraphQLEndpoint } from './verifyGraphQLEndpoint.js';

/**
 * Ensure that everything Faust requires to run is available.
 */
export async function healthCheck(): Promise<void> {
  // Check Faust Env variables before continuing.
  await validateFaustEnvVars();

  // Perform our health checks.
  await verifyGraphQLEndpoint();
}
