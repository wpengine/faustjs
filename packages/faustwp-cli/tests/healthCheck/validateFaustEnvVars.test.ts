import { validateFaustEnvVars } from '../../src/healthCheck/validateFaustEnvVars';
import fetchMock from 'fetch-mock';
/**
 * @jest-environment jsdom
 */
describe('healthCheck/validateFaustEnvVars', () => {
  const envBackup = process.env;

  beforeEach(() => {
    process.env = { ...envBackup };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    process.env = envBackup;
  });

  it('exits with a 1 exit code when the WordPress URL is undefined', async () => {
    // @ts-ignore
    const mockExit = jest.spyOn(process, 'exit').mockImplementation((code) => {
      if (code && code !== 0) {
        throw new Error(`Exit code: ${code}`);
      }
    });

    // Use try/catch block to mock process.exit
    try {
      await validateFaustEnvVars();
    } catch (err) {
      console.log(err);
    }

    expect(mockExit).toHaveBeenCalledWith(1);
  });

  it('does not exit or throw an error when the WordPress URL is set', async () => {
    // @ts-ignore
    const mockExit = jest.spyOn(process, 'exit').mockImplementation((code) => {
      if (code && code !== 0) {
        return;
      }
    });

    process.env.NEXT_PUBLIC_WORDPRESS_URL = 'http://headless.local';

    await validateFaustEnvVars();

    expect(mockExit).toBeCalledTimes(0);
  });

  it('logs an error when the secret key validation fails', async () => {

    process.env.NEXT_PUBLIC_WORDPRESS_URL = 'https://headless.local';
    process.env.FAUST_SECRET_KEY = 'invalid-secret-key';

    fetchMock.post('https://headless.local/wp-json/faustwp/v1/validate_secret_key', {
      status: 401,
    });
    
    await validateFaustEnvVars();

    return expect(Promise.resolve(validateFaustEnvVars())).toMatchSnapshot('Check to ensure your FAUST_SECRET_KEY matches your Faust Secret Key under wp-admin settings');
  });

});
