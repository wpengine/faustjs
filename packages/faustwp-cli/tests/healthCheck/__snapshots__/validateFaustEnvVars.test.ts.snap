// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`healthCheck/validateFaustEnvVars logs an error when the secret key validation fails: Ensure your FAUST_SECRET_KEY environment variable matches your Secret Key in the Faust WordPress plugin settings 1`] = `Promise {}`;
